azkaban-common/src/main/java/azkaban/scheduler/MissedSchedulesManager.java

```diff
@@ -0,0 +1,255 @@
+  private final long DEFAULT_THREAD_IDLE_TIME_IN_MS = 60000;
+  private final int threadPoolSize;
+  private final int DEFAULT_THREAD_POOL_SIZE = 50;
+
+  @Inject
+  public MissedSchedulesManager(final Props azkProps,
+      final ProjectManager projectManager,
+      final Emailer emailer) {
+    this.projectManager = projectManager;
+    this.emailer = emailer;
+    this.sleepIntervalInMs = azkProps.getLong(Constants.MISSED_SCHEDULE_THREAD_IDLE_TIME, DEFAULT_THREAD_IDLE_TIME_IN_MS);
+    this.threadPoolSize = azkProps.getInt(Constants.MISSED_SCHEDULE_THREAD_POOL_SIZE, DEFAULT_THREAD_POOL_SIZE);
+    if (!azkProps.getBoolean(Constants.MISSED_SCHEDULE_MANAGER_ENABLED, false) || threadPoolSize <= 0) {
+      if (threadPoolSize <= 0) {
+        LOG.warn("An invalid thread pool size is passed in: " + threadPoolSize + ", assuming that the service "
+            + "is disabled, will not start any thread to do missed schedules operations.");
+      }
+      this.missedSchedulesExecutor = null;
+      this.backgroundTasksList = null;
+      return;
+    }
+    this.missedSchedulesExecutor =
+        Executors.newFixedThreadPool(threadPoolSize, new DaemonThreadFactory("MissedSchedulesTask_thread"));
+    this.backgroundTasksList = new ArrayList<>(threadPoolSize);
+  }
+
+  /**
+   * Put timestamps for Missed executions along with email notifications into Task Queue.
+   *
+   * @param missedScheduleTimesInMs, timestamps of missed schedule
+   * @param action, execute flow action
+   * @param backExecutionEnabled, a schedule config from user, the default is false
+   * */
+  public boolean addMissedSchedule(final List<Long> missedScheduleTimesInMs,
+      final ExecuteFlowAction action,
+      final boolean backExecutionEnabled) {
+    String projectId = action.getProjectName();
+    String flowName = action.getFlowName();
+    Project project = projectManager.getProject(projectId);
+    if (project == null) {
+      return false;
+    }
+    if (backExecutionEnabled) {
+      return this.missedScheduleTaskQueue.offer(
+          new MissedScheduleTaskQueueNode(
+              missedScheduleTimesInMs,
+              project.getFlow(flowName).getFailureEmails(),
+              action)
+      );
+    }
+    return this.missedScheduleTaskQueue.offer(
+        new MissedScheduleTaskQueueNode(
+            missedScheduleTimesInMs,
+            project.getFlow(flowName).getFailureEmails(),
+            projectId,
+            flowName));
+  }
+
+  @VisibleForTesting
+  protected boolean isTaskQueueEmpty() {
+    return this.missedScheduleTaskQueue.isEmpty();
+  }
+
+  @VisibleForTesting
+  protected MissedScheduleTaskQueueNode peekFirstTask() {
+    return this.missedScheduleTaskQueue.peek();
+  }
+
+  public void start() {
+    // Start the thread pool
+    for (int i = 0; i < threadPoolSize; i++) {
+      MissedSchedulesOperationTask task = new MissedSchedulesOperationTask(i, sleepIntervalInMs);
+      this.backgroundTasksList.add(task);
```

Comment:
If MISSED_SCHEDULE_MANAGER_ENABLED is false, backgroundTasksList is null. Will throw NPE here.